### PR TITLE
Metadata

### PR DESCRIPTION
**Pull request by DTadrzak**
Add: neutron-metadata-agents
Change: split neutron-agents into separated deployments.

### CODE REVIEW

File: neutron/templates/agents/metadata-agent.yaml.j2 [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    spec:
+      nodeSelector:
+        {{ deployment.control_node_label }}: enabled
+      securityContext:
+        runAsUser: 0
+      dnsPolicy: ClusterFirst
+      hostNetwork: true
+      containers:
+        - name: neutron-metadata-agent
+          image: {{ deployment.image.metadata }}
```
Comment (Quentin-M) [inline neutron/templates/agents/metadata-agent.yaml.j2]: Add the variable in the manifest

File: neutron/templates/configmaps/metadata-agent.ini.yaml.j2 [hunk 2]
```diff
@@ -3,6 +3,30 @@ kind: ConfigMap
...
+    project_domain_id = default
+    user_domain_id = default
+    project_name = service
+    username = {{ keystone.admin_user }}
+    password = {{ keystone.admin_password }}
+    endpoint_type = adminURL
+
+    # Nova metadata service IP and port
+    nova_metadata_ip = {{ network.ip_address }}
     nova_metadata_port = {{ metadata.port }}
```
Comment (Quentin-M) [inline neutron/templates/configmaps/metadata-agent.ini.yaml.j2]: Rename to `network.port.metadata` and add to manifest.

File: neutron/templates/configmaps/metadata-agent.ini.yaml.j2 [hunk 3]
```diff
@@ -3,6 +3,30 @@ kind: ConfigMap
...
+
+    # Metadata proxy shared secret
     metadata_proxy_shared_secret = {{ metadata.secret }}
```
Comment (Quentin-M) [inline neutron/templates/configmaps/metadata-agent.ini.yaml.j2]: Rename to `neutron.metadata_secret`, add to manifest with default and be sure to duplicate the variable (and its default) into the nova's manifest as well and use it in nova.conf.

File: neutron/templates/agents/metadata-agent.yaml.j2 (hunk 1 above)
Comment (DTadrzak) [inline neutron/templates/agents/metadata-agent.yaml.j2]: Done

File: neutron/templates/configmaps/metadata-agent.ini.yaml.j2 (hunk 2 above)
Comment (DTadrzak) [inline neutron/templates/configmaps/metadata-agent.ini.yaml.j2]: Done

File: neutron/templates/configmaps/metadata-agent.ini.yaml.j2 (hunk 3 above)
Comment (DTadrzak) [inline neutron/templates/configmaps/metadata-agent.ini.yaml.j2]: Done

File: neutron/templates/configmaps/neutron.conf.yaml.j2 [hunk 4]
```diff
@@ -3,10 +3,11 @@ kind: ConfigMap
 data:
   neutron.conf: |+
     [DEFAULT]
-    debug = {{ misc.debug }}
+    debug = True
```
Comment (Quentin-M) [inline neutron/templates/configmaps/neutron.conf.yaml.j2]: Leftover?

File: neutron/manifest.jsonnet [hunk 5]
```diff
@@ -38,6 +39,10 @@ kpm.package({
       ip_address: "{{ .IP }}",
       minion_interface_name: "eno1",
 
+      nova: {
+        ip: "nova-api"  ,
```
Comment (Quentin-M) [inline neutron/manifest.jsonnet]: nitpicking: we usually name these "address" in the other packages.

**Comment by Quentin-M** [comment]
I'm not reviewing the whole thing once again, made two small comments, other than that, it looks great overall: LGTM!

**Comment by DTadrzak** [comment]
@ss7pro 
@Quentin-M 
@PiotrProkop 
@mateusz-blaszkowski

**Comment by PiotrProkop** [comment]
Please make sure neutron-metadata-agent and neutron-l3-agent are scheduled on the same node.

File: neutron/templates/agents/metadata-agent.yaml.j2 [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+              mountPath: /configmaps/metadata-agent.ini
+            - name: resolvconf
+              mountPath: /configmaps/resolv.conf
+            - name: libmodules
+              mountPath: /lib/modules
+              readOnly: true
+            - name: runopenvswitch
+              mountPath: /run/openvswitch
+            - name: usrsbinip
+              mountPath: /usr/sbin/ip
```
Comment (ss7pro) [inline neutron/templates/agents/metadata-agent.yaml.j2]: remove that.

File: neutron/templates/agents/metadata-agent.yaml.j2 [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+            - name: neutronconf
+              mountPath: /configmaps/neutron.conf
+            - name: ml2confini
+              mountPath: /configmaps/ml2-conf.ini
+            - name: metadataagentini
+              mountPath: /configmaps/metadata-agent.ini
+            - name: resolvconf
+              mountPath: /configmaps/resolv.conf
+            - name: libmodules
+              mountPath: /lib/modules
```
Comment (ss7pro) [inline neutron/templates/agents/metadata-agent.yaml.j2]: remove that.

File: neutron/templates/agents/metadata-agent.yaml.j2 [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+              readOnly: true
+            - name: runopenvswitch
+              mountPath: /run/openvswitch
+            - name: usrsbinip
```
Comment (ss7pro) [inline neutron/templates/agents/metadata-agent.yaml.j2]: remove that as well.

File: neutron/templates/agents/metadata-agent.yaml.j2 [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+              mountPath: /usr/sbin/ip
+            - name: socket
+              mountPath: /var/lib/neutron/kolla/
```
Comment (ss7pro) [inline neutron/templates/agents/metadata-agent.yaml.j2]: rename it to simply /var/lib/neutron.

File: neutron/templates/agents/metadata-agent.yaml.j2 [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+            path: /lib/modules
+        - name: runopenvswitch
+          hostPath:
+            path: /run/openvswitch
+        - name: usrsbinip
+          hostPath:
+            path: /usr/bin/ip
+        - name: socket
+          hostPath:
+            path: /var/lib/neutron/kolla/
```
Comment (ss7pro) [inline neutron/templates/agents/metadata-agent.yaml.j2]: rename to /var/lib/neutron.

File: neutron/templates/agents/metadata-agent.yaml.j2 [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+        - name: resolvconf
+          configMap:
+            name: neutron-resolvconf
+        - name: libmodules
+          hostPath:
+            path: /lib/modules
+        - name: runopenvswitch
+          hostPath:
+            path: /run/openvswitch
+        - name: usrsbinip
```
Comment (ss7pro) [inline neutron/templates/agents/metadata-agent.yaml.j2]: remove.

File: neutron/templates/agents/metadata-agent.yaml.j2 [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+        - name: ml2confini
+          configMap:
+            name: neutron-ml2confini
+        - name: metadataagentini
+          configMap:
+            name: neutron-metadataagentini
+        - name: resolvconf
+          configMap:
+            name: neutron-resolvconf
+        - name: libmodules
```
Comment (ss7pro) [inline neutron/templates/agents/metadata-agent.yaml.j2]: remove.

File: neutron/templates/configmaps/dnsmasq.conf.yaml.j2 [hunk 13]
```diff
@@ -2,4 +2,6 @@ apiVersion: v1
 kind: ConfigMap
 data:
   dnsmasq.conf: |+
-    log-facility=-
+    log-facility=/var/log/dnsmasq.log
```
Comment (ss7pro) [inline neutron/templates/configmaps/dnsmasq.conf.yaml.j2]: log files is not needed.

File: neutron/templates/configmaps/neutron.conf.yaml.j2 [hunk 14]
```diff
@@ -22,11 +23,15 @@ data:
...
 
-    allow_overlapping_ips = true
+    allow_overlapping_ips = True
     core_plugin = ml2
     service_plugins = router
 
-    interface_driver = neutron.agent.linux.interface.OVSInterfaceDriver
+    interface_driver = openvswitch
+
+    metadata_proxy_socket = /var/lib/neutron/kolla/metadata_proxy
```
Comment (ss7pro) [inline neutron/templates/configmaps/neutron.conf.yaml.j2]: remove kolla from the path.

File: neutron/templates/agents/l3-agent.yaml.j2 [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+              mountPath: /configmaps/resolv.conf
+            - name: libmodules
+              mountPath: /lib/modules
+              readOnly: true
+            - name: runopenvswitch
+              mountPath: /run/openvswitch
+            - name: usrsbinip
+              mountPath: /usr/sbin/ip
+            - name: socket
+              mountPath: /var/lib/neutron/kolla/
```
Comment (ss7pro) [inline neutron/templates/agents/l3-agent.yaml.j2]: remove kolla from path.

File: neutron/templates/agents/l3-agent.yaml.j2 [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+              mountPath: /configmaps/l3-agent.ini
+            - name: resolvconf
+              mountPath: /configmaps/resolv.conf
+            - name: libmodules
+              mountPath: /lib/modules
+              readOnly: true
+            - name: runopenvswitch
+              mountPath: /run/openvswitch
+            - name: usrsbinip
+              mountPath: /usr/sbin/ip
```
Comment (ss7pro) [inline neutron/templates/agents/l3-agent.yaml.j2]: remove

File: neutron/templates/agents/l3-agent.yaml.j2 [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+            - name: neutronconf
+              mountPath: /configmaps/neutron.conf
+            - name: ml2confini
+              mountPath: /configmaps/ml2-conf.ini
+            - name: l3agentini
+              mountPath: /configmaps/l3-agent.ini
+            - name: resolvconf
+              mountPath: /configmaps/resolv.conf
+            - name: libmodules
+              mountPath: /lib/modules
```
Comment (ss7pro) [inline neutron/templates/agents/l3-agent.yaml.j2]: remove.

File: neutron/templates/agents/l3-agent.yaml.j2 [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+        - name: l3agentini
+          configMap:
+            name: neutron-l3agentini
+        - name: libmodules
+          hostPath:
+            path: /lib/modules
+        - name: runopenvswitch
+          hostPath:
+            path: /run/openvswitch
+        - name: usrsbinip
```
Comment (ss7pro) [inline neutron/templates/agents/l3-agent.yaml.j2]: remove.

File: neutron/templates/agents/l3-agent.yaml.j2 [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+        - name: ml2confini
+          configMap:
+            name: neutron-ml2confini
+        - name: resolvconf
+          configMap:
+            name: neutron-resolvconf
+        - name: l3agentini
+          configMap:
+            name: neutron-l3agentini
+        - name: libmodules
```
Comment (ss7pro) [inline neutron/templates/agents/l3-agent.yaml.j2]: remove.

File: neutron/templates/agents/l3-agent.yaml.j2 [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+            path: /lib/modules
+        - name: runopenvswitch
+          hostPath:
+            path: /run/openvswitch
+        - name: usrsbinip
+          hostPath:
+            path: /usr/bin/ip
+        - name: socket
+          hostPath:
+            path: /var/lib/neutron/kolla/
```
Comment (ss7pro) [inline neutron/templates/agents/l3-agent.yaml.j2]: remove kolla from path.

File: neutron/templates/agents/metadata-agent.yaml.j2 [hunk 21]
```diff
@@ -0,0 +1,95 @@
+apiVersion: extensions/v1beta1
+kind: Deployment
```
Comment (ss7pro) [inline neutron/templates/agents/metadata-agent.yaml.j2]: for now this needs to be a daemonSet.

**Comment by PiotrProkop** [comment]
LGTM.

**Pull request merged**